audmath/core/api.py

```diff
@@ -677,6 +677,79 @@ def samples(
     return round(duration * sampling_rate)
 
 
+def similarity(
+        u: typing.Union[typing.Sequence, np.ndarray],
+        v: typing.Union[typing.Sequence, np.ndarray],
+) -> typing.Union[np.floating, np.ndarray]:
+    r"""Cosine similarity between two arrays.
+
+    If the incoming arrays are of size
+    :math:`(1, k)` or :math:`(k,)`,
+    a single similarity value is returned.
+    If the arrays are of size
+    :math:`(n, k)`
+    and :math:`(m, k)`
+    with :math:`n > 1`
+    or :math:`m > 1`,
+    an array of size
+    :math:`(n, m)`
+    with similarities is returned.
+
+    The input arrays can also be provided as
+    :class:`pandas.DataFrame`
+    or :class:`pandas.Series`.
+
+    The cosine similarity is given by
+    :math:`\frac{u \cdot v}{\lVert u\rVert_2 \lVert v\rVert_2}`.
+
+    Args:
+        u: input array
+        v: input array
+
+    Returns:
+        similarity between arrays
+
+    Example:
+        >>> similarity([1, 0], [1, 0])
+        1.0
+        >>> similarity([1, 0], [0, 1])
+        0.0
+        >>> similarity([1, 0], [-1, 0])
+        -1.0
+        >>> similarity([1, 0], [[1, 0], [0, 1]])
+        array([1., 0.])
+        >>> similarity([[1, 0], [0, 1]], [[1, 0], [0, 1]])
+        array([[1., 0.],
+               [0., 1.]])
+        >>> similarity([[1, 0], [0, 1]], [[1, 0], [0, 1], [-1, 0]])
+        array([[ 1.,  0., -1.],
+               [ 0.,  1.,  0.]])
+
+    """
+    def to_numpy(x):
+        if not isinstance(u, np.ndarray):
+            try:
+                # pandas object
+                x = x.to_numpy()
+            except AttributeError:
+                # sequence
+                x = np.array(x)
+        return np.atleast_2d(x)
+
+    u = to_numpy(u)
+    v = to_numpy(v)
+
+    u = u / np.linalg.norm(u, ord=2, keepdims=True, axis=-1)
+    v = v / np.linalg.norm(v, ord=2, keepdims=True, axis=-1)
+    sim = np.inner(u, v).squeeze()
+
+    # Ensure single value is not returned as array
+    if not sim.shape:
+        sim = np.float64(sim)
```

Comment:
I am wondering if it is better to return the single value also as array . I am thinking of a case where we compare 1 ... n elements with another element (sometimes its 1 element sometimes its n elements) and than e.g. calculate the mean. Then it would be a lot easier if the single value is also returned as an array.

Reply:
The current convention follows other implementations, e.g. `audmath.rms()` where single values are also returned as `float`, which also feels more convenient if you don't consider the case you are describing.

But I also see the need to handle your example out of the box.
One solution might be to handle `(1, k)` input differently than `(k,)`, e.g.

```python
>>> audmath.similarity([1, 0], [1, 0])
1.0

>>> audmath.similarity([[1, 0]], [1, 0])
array([1.0])
```

Another solution might to add a `keepdims` argument as we did in `audmath.rms()`.
It's not completely obvious what the behavior should be then, but maybe this:

```python
>>> audmath.similarity([1, 0], [1, 0], keepdims=True)
array([1.0])

>>> audmath.similarity([[1, 0]], [1, 0], keepdims=True)
array([[1.0]])
```

Any preference, or another idea?

Reply:
If we don't want to make the behavior dependent on the input, the following might also be an option:

```python
>>> audmath.similarity([1, 0], [1, 0], as_array=True)
array([1.0])

>>> audmath.similarity([[1, 0]], [1, 0], as_array=True)
array([1.0])
```

Reply:
I implemented the first solution, added more tests and updated the docstring.

Please read carefully through the tests and see if it behaves now for every test case as you would expect.

Reply:
Thank you for the adjustment!
It got a lot more complicated than I anticipated.
I am still twisting my head around what the implications might be of creating a 1d array when an input array is 1d.
But I guess as those are the cases, where "atleast_2d " adds dimensions it all works as expected.